lib/cloud_profiler_agent/pprof_builder.rb

```diff
@@ -0,0 +1,181 @@
+# frozen_string_literal: true
+
+require 'zlib'
+require 'stackprof'
+
+module CloudProfilerAgent
+  # PprofBuilder converts from stackprof to pprof formats.
+  #
+  # Typical usage:
+  #
+  #   start_time = Time.now
+  #   stackprof = StackProf.run(mode: :cpu, raw: true) do
+  #     # ...
+  #   end
+  #   pprof = PprofBuilder.convert_stackprof(stackprof, start_time, Time.now)
+  #   IO::binwrite('profile.pprof', pprof)
+  #
+  # StackProf must be invoked with raw: true for the conversion to work.
+  #
+  # The pprof format is a gzip-compressed protobuf, documented here:
+  # https://github.com/google/pprof/blob/master/proto/profile.proto
+  #
+  # The conversion is not quite isomorphic. In particular, each sample in pprof
+  # consists of a stack with line numbers and even instruction addresses.
+  # StackProf on the other hand records stack frames with only function-level
+  # granularity. Thus, the line numbers in the pprof output only reflect the
+  # first line of the function, regardless of which line was actually in the
+  # stack frame.
+  class PprofBuilder
+    def self.convert_stackprof(stackprof, start_time, end_time)
+      converter = PprofBuilder.new(stackprof, start_time, end_time)
+      converter.pprof_bytes
+    end
+
+    def initialize(profile, start_time, end_time)
+      @profile = profile
+      @start_time = start_time
+      @duration = end_time - start_time
+
+      @string_map = StringMap.new
+    end
+
+    # message returns a Perftools::Profiles::Profile, the deserialized version
+    # of a pprof profile.
+    def message
+      main_mapping = Perftools::Profiles::Mapping.new(
+        id: 1,
+        filename: @string_map.add('TODO')
+      )
+
+      message = Perftools::Profiles::Profile.new(
+        sample_type: [sample_type],
+        mapping: [main_mapping],
+        time_nanos: @start_time.to_i * 1_000_000_000,
+        duration_nanos: @duration * 1_000_000_000,
+        period_type: sample_type,
+        period: period,
+        default_sample_type: 0
+      )
+      process_raw(message)
+      process_frames(message)
+      message.string_table += @string_map.strings
+      message
+    end
+
+    # pprof_bytes returns a gzip'd protobuf object, like would be written to
+    # disk, returned by a profiling endpoint, or otherwise consumed by the
+    # `pprof` tool.
+    def pprof_bytes
+      Zlib.gzip(Perftools::Profiles::Profile.encode(message))
+    end
+
+    private
+
+    def to_pprof_unit(value)
+      case @profile[:mode]
```

Comment:
Use `fetch` here and everywhere below that you fetch a key from `@profile`.
```suggestion
      case @profile.fetch(:mode)
```